src/app/utils/xml-utils.ts

```diff
@@ -36,3 +36,141 @@ export function replaceNotWordChar(textContent: string) {
+  };
+  const evtTextComplexElements = ['choice', 'app', 'l', 'quote', 'p', 'lg'];
+  let out = '';
+  elem.childNodes.forEach((x) => (evtTextElements[x.nodeName] !== undefined) ? out += x[ evtTextElements[x.nodeName] ] : (
+    evtTextComplexElements.includes(x.nodeName) ? out += chainDeepTexts(x) : '' ))
+
+  return out;
+}
+
+/**
+ * Retrieve and chain textContent of all descendents
+ * @param elem ChildNode
+ * @returns out string
+ */
+export function chainDeepTexts(elem: ChildNode): string {
+  const evtInnerTextElements = ['#text', 'reg', 'corr', 'rdg'];
+  const textProperty = 'textContent';
+  let out = '';
+  elem.childNodes.forEach((x) => (evtInnerTextElements.includes(x.nodeName)) ? ((x[textProperty] !== null) ? out += x[textProperty] : '') : '')
+
+  return out;
+}
+
+/**
+* Retrieve external elements searching between provided elements types and filtering on attributes base
+* It searches all document for a bibl with the correct xml:id.
+* It would be faster if we knew the id or unique element to search in
+* @param element XMLElement
+* @param attrSourceNames
+* @param attrTargetName
+* @returns array of XMLElement
+*/
+export function getExternalElements(elem: XMLElement, attrSourceNames: string[], attrTargetName: string, elTypes: string): XMLElement[] {
+  const sourceIDs = attrSourceNames.map((x) => elem.getAttribute(x));
+  const sourcesToFind = sourceIDs.filter((x) => x).map((x) => x.replace('#',''));
+
+  if (sourcesToFind.length === 0) {
+    return [];
+  }
+
+  return Array.from(elem.ownerDocument.querySelectorAll<XMLElement>(elTypes))
+    .filter((x) => sourcesToFind.includes(x.getAttribute(attrTargetName)))
+}
+
+export function isAnalogue(elem: XMLElement, markerAttrs: string[]): boolean {
+  return (markerAttrs.includes(elem.getAttribute('type')));
+}
+
+export function isSource(elem: XMLElement, attrs: string[]): boolean {
+  let validAttrs = false;
+  attrs.forEach((attr) => { if (elem.getAttribute(attr) !== null) { validAttrs = true } });
+
+  return (validAttrs);
+}
+
+/**
+ * Get an element of the same type with the given xml:id and update it with a copy of the source element
+ * @param fromElement XMLElement
+ * @param toXMLID string
+ */
+export function getCorrespElement(fromElement: XMLElement, toXMLID: string) {
+  const findID = toXMLID.replace('#','');
+  const sameTypeElements = fromElement.ownerDocument.querySelectorAll<XMLElement>(fromElement.tagName);
+  let found = null;
+  sameTypeElements.forEach((x) => { if (x.getAttribute('xml:id') === findID) { found = x; return }});
+  if (found !== null) {
+    found.appendChild(fromElement.cloneNode());
+  }
+}
+
+
+/**
+ * Retrieve textContent and elements between the provided element and the one found with the xml:id provided
+ * @param fromElement XMLElement
+ * @param toXMLID string
+ * @returns object { text: string, elements: any }
+ */
+export function getSpanToElements(fromElement: XMLElement, toXMLID: string): { text: string, elements: any } {
```

Comment:
Then name of the function does not properly explain what it does. Better something like `getContentBetweenElementAndId`